src/theory/strings/core_solver.cpp

```diff
@@ -1954,47 +2106,38 @@ void CoreSolver::processDeq(Node ni, Node nj)
         // are both non-constants. We split them into parts that have the same
         // lengths.
         //
-        // E.g. x ++ x' ++ ... != y ++ y' ++ ... ^ len(x) != len(y) --->
-        //      len(k1) = len(x) ^ len(k2) = len(y) ^
-        //        (y = k1 ++ k3 v x = k1 ++ k2)
-        Trace("strings-solve") << "Non-Simple Case 1 : add lemma " << std::endl;
-        std::vector<Node> antec(nfni.d_exp.begin(), nfni.d_exp.end());
-        antec.insert(antec.end(), nfnj.d_exp.begin(), nfnj.d_exp.end());
-        std::vector<Node> antecNewLits;
-
-        if (d_state.areDisequal(ni, nj))
-        {
-          antec.push_back(ni.eqNode(nj).negate());
-        }
-        else
+        // len(x) >= len(y) => x = k1 ++ k2 ^ len(k1) = len(y)
+        // len(y) >= len(x) => y = k3 ++ k4 ^ len(k3) = len(x)
+        Trace("strings-solve")
+            << "Non-Simple Case 1 : add lemmas " << std::endl;
+        SkolemCache* skc = d_termReg.getSkolemCache();
+
+        for (unsigned r = 0; r < 2; r++)
         {
-          antecNewLits.push_back(ni.eqNode(nj).negate());
+          Node ux = r == 0 ? x : y;
+          Node uxLen = nm->mkNode(STRING_LENGTH, x);
+          Node uyLen = nm->mkNode(STRING_LENGTH, y);
```

Comment:
I believe that there is a bug here: If `r == 1`, then we decompose `y` using `len(y)`, instead of `len(x)`.

```suggestion
          Node ux = r == 0 ? x : y;
          Node uxLen = nm->mkNode(STRING_LENGTH, r == 0 ? x : y);
          Node uyLen = nm->mkNode(STRING_LENGTH, r == 0 ? y : x);
```

Reply:
Yes you're right, good catch.